File: 542-01Matrix/542-01Matrix.java
/*

542. 01 Matrix
Medium

Given a matrix consists of 0 and 1, find the distance of the nearest 0 for each cell.

The distance between two adjacent cells is 1.

Example 1:

Input:
[[0,0,0],
 [0,1,0],
 [0,0,0]]

Output:
[[0,0,0],
 [0,1,0],
 [0,0,0]]

Example 2:

Input:
[[0,0,0],
 [0,1,0],
 [1,1,1]]

Output:
[[0,0,0],
 [0,1,0],
 [1,2,1]]

Note:

    The number of elements of the given matrix will not exceed 10,000.
    There are at least one 0 in the given matrix.
    The cells are adjacent in only four directions: up, down, left and right.

*/

class Solution {
    public int[][] updateMatrix(int[][] matrix) {
        //Apprach BFS
        int[][] res = new int[matrix.length][matrix[0].length];
        boolean[][] vis = new boolean[matrix.length][matrix[0].length];
        int row = matrix.length;
        int col = matrix[0].length;
        Queue<int[]> q = new LinkedList<>();
        for (int i = 0; i<row; i++){
            for (int j = 0; j<col; j++){
                if (matrix[i][j] == 0){
                    q.add(new int[]{i,j});
                    vis[i][j] = true;
                }
            }
        }
        int dirs[][] = new int[][]{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        while(!q.isEmpty()){
                int temp[] = q.poll();
                for(int[]dir : dirs){
                    int x = temp[0] + dir[0];
                    int y = temp[1] + dir[1];
                    if (x < 0 || y < 0 || x >= row || y >= col || vis[x][y]) continue;
                    vis[x][y] = true;
                    res[x][y] = res[temp[0]][temp[1]] + 1;
                    q.add(new int[]{x, y});
                }
        }
        return res; 
    }
}

File: 542-01Matrix/542-01Matrix.py
"""

542. 01 Matrix

Given an m x n binary matrix mat, return the distance of the nearest 0 for each cell.

The distance between two cells sharing a common edge is 1.

Example 1:

Input: mat = [[0,0,0],[0,1,0],[0,0,0]]
Output: [[0,0,0],[0,1,0],[0,0,0]]

Example 2:

Input: mat = [[0,0,0],[0,1,0],[1,1,1]]
Output: [[0,0,0],[0,1,0],[1,2,1]]

Constraints:

    m == mat.length
    n == mat[i].length
    1 <= m, n <= 104
    1 <= m * n <= 104
    mat[i][j] is either 0 or 1.
    There is at least one 0 in mat.

Note: This question is the same as 1765: https://leetcode.com/problems/map-of-highest-peak/

"""

class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        q = deque()
        vis = [[False] * len(mat[0]) for _ in range(len(mat))]
        for i in range(len(mat)):
            for j in range(len(mat[0])):
                if mat[i][j] == 0:
                    q.append([i, j])
                    vis[i][j] = True
        dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]]
        res = 1
        while q:
            lenn = len(q)
            for i in range(lenn):
                curr = q.popleft()
                for d in dirs:
                    x = curr[0] + d[0]
                    y = curr[1] + d[1]
                    if x < 0 or y < 0 or x >= len(mat) or y >= len(mat[0]) or vis[x][y]: continue
                    mat[x][y] = res
                    vis[x][y] = True
                    q.append([x, y])
            res += 1
        return mat
